// mongodb-mcp-server/dist/cjs/tools/mongodb/read/count.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CountTool = exports.CountArgs = void 0;
const mongodbTool_js_1 = require("../mongodbTool.js");
const indexCheck_js_1 = require("../../../helpers/indexCheck.js");
const args_js_1 = require("../../args.js");
exports.CountArgs = {
    query: (0, args_js_1.zEJSON)()
        .optional()
        .describe("A filter/query parameter. Allows users to filter the documents to count. Matches the syntax of the filter argument of db.collection.count()."),
};
class CountTool extends mongodbTool_js_1.MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.description = "Gets the number of documents in a MongoDB collection using db.collection.count() and query as an optional filter parameter";
        this.argsShape = {
            ...mongodbTool_js_1.DbOperationArgs,
            ...exports.CountArgs,
        };
    }
    async execute({ database, collection, query }, { signal }) {
        const provider = await this.ensureConnected();
        // Check if count operation uses an index if enabled
        if (this.config.indexCheck) {
            await (0, indexCheck_js_1.checkIndexUsage)({
                database,
                collection,
                operation: "count",
                explainCallback: async () => {
                    return provider.runCommandWithCheck(database, {
                        explain: {
                            count: collection,
                            query,
                        },
                        verbosity: "queryPlanner",
                    }, {
                        signal,
                    });
                },
                logger: this.session.logger,
            });
        }
        const count = await provider.countDocuments(database, collection, query, {
            // @ts-expect-error signal is available in the driver but not NodeDriverServiceProvider MONGOSH-3142
            signal,
        });
        return {
            content: [
                {
                    text: `Found ${count} documents in the collection "${collection}"${query ? " that matched the query" : ""}.`,
                    type: "text",
                },
            ],
        };
    }
}
exports.CountTool = CountTool;
CountTool.toolName = "count";
CountTool.operationType = "read";
